src/js/plugins/ethereum/typedData.js

```diff
@@ -0,0 +1,51 @@
+// eslint-disable-next-line import/no-unresolved
+const sigUtil = require('@metamask/eth-sig-util');
+
+// Sanitization is used for T1 as eth-sig-util does not support BigInt
+function sanitizeData(data) {
+    switch (Object.prototype.toString.call(data)) {
+        case '[object Object]':
+            const entries = Object.keys(data).map(k => [k, sanitizeData(data[k])]);
+            return Object.fromEntries(entries);
+
+        case '[object Array]':
+            return data.map(v => sanitizeData(v));
+
+        case '[object BigInt]':
+            return data.toString();
+
+        default:
+            return data;
+    }
+}
+
+const transformTypedData = (data, metamask_v4_compat) => {
+    const { types, primaryType, domain, message } = sigUtil.TypedDataUtils.sanitizeData(data);
+
+    const version = metamask_v4_compat
+        ? sigUtil.SignTypedDataVersion.V4
+        : sigUtil.SignTypedDataVersion.V3;
```

Comment:
this is wrong -- `metamask_v4_compat=False` is in fact  metamask V5 (doesn't exist yet, i believe)

instead we should always use `V4` and always set `metamask_v4_compat = True` if this is being used.

Reply:
I believe `metamask_v4_compat` comes from the developers and they might send `false` here (I don't know why they would but they could).

Ledger already raises an exception if that is set to false. We can also do the same thing. Do you agree?

Reply:
Yes, that is a good idea.

Reply:
Fixed in 33d8b9007fd1625a3cd283bf38aa90adea43d737